File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
from collections import deque
input= sys.stdin.readline
sys.setrecursionlimit(2600)

N,M = map(int,input().split())

graph = [[] for _ in range(N+1)]

visit = [False]*(N+1)


for _ in range(M):
    v1,v2 = map(int,input().split())
    graph[v1].append(v2)
    graph[v2].append(v1)

def bfs(start):
    queue = deque([start])
    visit[start] = True
    while queue:
        v = queue.popleft()
        for i in graph[v]:
            if not visit[i]:
                queue.append(i)
                visit[i]=True

cnt = 0
for i in range(1,N+1):
    if not visit[i]: 
        bfs(i)
        cnt+=1
print(cnt)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static int n,m;
	static boolean[] visit;
	static int cnt = 0;
	static HashMap<Integer, List<Integer>> map = new HashMap<>();
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		visit = new boolean[n + 1];
		for (int i = 1; i <= n; i++) {
			map.put(i, new ArrayList<>());
		}
		
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());
			int num1 = Integer.parseInt(st.nextToken());
			int num2 = Integer.parseInt(st.nextToken());
			map.get(num1).add(num2);
			map.get(num2).add(num1);
		}

		for (int i = 1; i <= n ; i++) {
			if(!visit[i]) {
				visit[i] = true;
				bfs(i);
				cnt++;
			}
		}
		
		System.out.println(cnt);
	}
	private static void bfs(int i) {
		Queue<Integer> q = new LinkedList<>();
		q.add(i);
		
		while(!q.isEmpty()) {
			int num = q.poll();
			
			List<Integer> tmp = map.get(num);
			for (int j = 0; j < tmp.size(); j++) {
				if(!visit[tmp.get(j)]) {
					visit[tmp.get(j)] = true;
					q.add(tmp.get(j));
				}
			}
		}
		
	}
}
